File: python/mirage/runtime/binding.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from __future__ import annotations

from collections.abc import Callable
from dataclasses import replace
from types import MappingProxyType

from mirage.ops.types import NamespaceView, SessionView
from mirage.process.types import ProcessView
from mirage.runtime.resolver import MountResolver
from mirage.runtime.types import DispatchFn, RuntimeContext
from mirage.types import PathSpec
from mirage.utils.context_scope import ContextScope


class WorkspaceBinding:
    """Local workspace connection shared by execution and filesystem adapters.

    Dispatch and routing stay live. Capture constructs per-execution views;
    holding the binding never holds a mutable current-session slot.
    """

    def __init__(
        self,
        dispatch: DispatchFn,
        resolver: MountResolver,
        context: Callable[[WorkspaceBinding], RuntimeContext] | None = None
    ) -> None:
        self.dispatch = dispatch
        self.resolver = resolver
        self._context = context

    def capture(self) -> RuntimeContext:
        return self._context(
            self) if self._context is not None else capture_binding(self)


def capture_binding(binding: WorkspaceBinding,
                    *,
                    ns: NamespaceView | None = None,
                    session_view: SessionView | None = None,
                    processes: ProcessView | None = None,
                    cwd: PathSpec | None = None,
                    env: dict[str, str] | None = None) -> RuntimeContext:
    """Capture callback context while reading live workspace state."""
    scope = ContextScope()
    view = ns or NamespaceView()
    links = view.links
    if links is not None:
        links = replace(links,
                        stat_at=scope.wrap(links.stat_at),
                        children=scope.wrap(links.children),
                        subtree=scope.wrap(links.subtree),
                        resolve=scope.wrap(links.resolve),
                        exists=scope.wrap_async(links.exists),
                        target_stat=scope.wrap_async(links.target_stat))
    mounts = view.mounts
    if mounts is not None:
        mounts = replace(mounts,
                         descendants=scope.wrap(mounts.descendants),
                         visible_descendants=scope.wrap(
                             mounts.visible_descendants),
                         is_root=scope.wrap(mounts.is_root),
                         root_of=scope.wrap(mounts.root_of))
    view = replace(view,
                   links=links,
                   mounts=mounts,
                   stat_overlay=scope.wrap(view.stat_overlay)
                   if view.stat_overlay else None,
                   child_mounts=scope.wrap(view.child_mounts)
                   if view.child_mounts else None)
    if session_view is not None:
        session_view = replace(session_view,
                               get=scope.wrap(session_view.get),
                               snapshot=scope.wrap(session_view.snapshot),
                               set=scope.wrap_async(session_view.set),
                               unset=scope.wrap_async(session_view.unset),
                               mark=scope.wrap_async(session_view.mark),
                               is_readonly=scope.wrap(
                                   session_view.is_readonly),
                               profile=scope.wrap(session_view.profile))
    resolver = ScopedResolver(binding.resolver, scope)
    return RuntimeContext(binding=binding,
                          dispatch=scope.wrap_async(binding.dispatch),
                          resolver=resolver,
                          ns=view,
                          session_view=session_view,
                          cwd=cwd or PathSpec.from_str_path("/"),
                          env=MappingProxyType(dict(env or {})),
                          scope=scope,
                          processes=processes)


class ScopedResolver:
    """Delegate every routing question under the captured context."""

    def __init__(self, resolver: MountResolver, scope: ContextScope) -> None:
        self._resolver = resolver
        self._scope = scope

    def prefixes(self) -> list[str]:
        return self._scope.call(self._resolver.prefixes)

    def owner_of(self, path: str) -> str | None:
        return self._scope.call(self._resolver.owner_of, path)

    def link_children(self, directory: str) -> set[str]:
        return self._scope.call(self._resolver.link_children, directory)

File: typescript/packages/core/src/runtime/binding.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import type { ProcessView } from '../process/types.ts'
import { captureSessionContext } from '../context/session_context.ts'
import { captureRecordingContext } from '../observe/context.ts'
import type { NamespaceView, SessionView } from '../ops/types.ts'
import { PathSpec } from '../types.ts'
import { ContextScope } from '../utils/context_scope.ts'
import type { MountResolver } from './resolver.ts'
import type { BridgeDispatchFn, RuntimeContext } from './types.ts'

/** Live workspace connection; capture creates views scoped to one execution. */
export class WorkspaceBinding {
  constructor(
    readonly dispatch: BridgeDispatchFn,
    readonly resolver: MountResolver,
    private readonly context?: (binding: WorkspaceBinding) => RuntimeContext,
  ) {}

  capture(): RuntimeContext {
    return this.context === undefined ? captureBinding(this) : this.context(this)
  }
}

/** Pin callback attribution without copying authoritative workspace state. */
export function captureBinding(
  binding: WorkspaceBinding,
  views: {
    ns?: NamespaceView
    sessionView?: SessionView
    processes?: ProcessView
    cwd?: PathSpec
    env?: Readonly<Record<string, string>>
  } = {},
  scope = new ContextScope([...captureSessionContext(), ...captureRecordingContext()]),
): RuntimeContext {
  const source = views.ns ?? {}
  const links = source.links
  const mounts = source.mounts
  const ns: NamespaceView = {
    ...source,
    ...(links === undefined
      ? {}
      : {
          links: {
            statAt: scope.wrap(links.statAt.bind(links)),
            children: scope.wrap(links.children.bind(links)),
            subtree: scope.wrap(links.subtree.bind(links)),
            resolve: scope.wrap(links.resolve.bind(links)),
            exists: scope.wrap(links.exists.bind(links)),
            targetStat: scope.wrap(links.targetStat.bind(links)),
          },
        }),
    ...(mounts === undefined
      ? {}
      : {
          mounts: {
            descendants: scope.wrap(mounts.descendants.bind(mounts)),
            visibleDescendants: scope.wrap(mounts.visibleDescendants.bind(mounts)),
            isRoot: scope.wrap(mounts.isRoot.bind(mounts)),
            rootOf: scope.wrap(mounts.rootOf.bind(mounts)),
          },
        }),
    ...(source.statOverlay === undefined ? {} : { statOverlay: scope.wrap(source.statOverlay) }),
    ...(source.childMounts === undefined ? {} : { childMounts: scope.wrap(source.childMounts) }),
  }
  const session = views.sessionView
  const sessionView =
    session === undefined
      ? null
      : {
          get: scope.wrap(session.get.bind(session)),
          snapshot: scope.wrap(session.snapshot.bind(session)),
          set: scope.wrap(session.set.bind(session)),
          unset: scope.wrap(session.unset.bind(session)),
          mark: scope.wrap(session.mark.bind(session)),
          isReadonly: scope.wrap(session.isReadonly.bind(session)),
          profile: scope.wrap(session.profile.bind(session)),
        }
  return Object.freeze({
    binding,
    dispatch: scope.wrap(binding.dispatch),
    resolver: {
      prefixes: scope.wrap(() => binding.resolver.prefixes()),
      ownerOf: scope.wrap((path: string) => binding.resolver.ownerOf(path)),
      linkChildren: scope.wrap((path: string) => binding.resolver.linkChildren(path)),
    },
    ns,
    sessionView,
    processes: views.processes ?? null,
    cwd: views.cwd ?? PathSpec.fromStrPath('/'),
    env: Object.freeze({ ...views.env }),
    scope,
  })
}
